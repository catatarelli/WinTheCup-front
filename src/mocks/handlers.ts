import { rest } from "msw";
import { REACT_APP_API_URL } from "@env";
import type {
  RegisterData,
  UserCredentials,
} from "../redux/features/user/userTypes";
import {
  mockGetPredictionByIdResponse,
  mockgetPredictionsResponse,
} from "./predictionsMocks";

export const handlers = [
  rest.post(`${REACT_APP_API_URL}/user/register`, async (req, res, ctx) => {
    const user = await req.json<RegisterData>();

    if (!user || user.email === "panchito@gmail.com") {
      return res(ctx.status(409));
    }

    return res(ctx.status(201), ctx.json({ user }));
  }),

  rest.post(`${REACT_APP_API_URL}/user/login`, async (req, res, ctx) => {
    const user = await req.json<UserCredentials>();
    if (user.password === "testPassword") {
      return res(ctx.status(401));
    }

    return res(ctx.status(200), ctx.json({ token: "token" }));
  }),

  rest.get(`${REACT_APP_API_URL}/predictions`, async (req, res, ctx) =>
    res.once(
      ctx.status(404),
      ctx.json({ error: "There was an error loading your predictions" })
    )
  ),

  rest.get(`${REACT_APP_API_URL}/predictions`, async (req, res, ctx) =>
    res.once(
      ctx.status(404),
      ctx.json({ error: "There was loading more predictions" })
    )
  ),

  rest.get(`${REACT_APP_API_URL}/predictions`, (req, res, ctx) =>
    res(ctx.status(200), ctx.json(mockgetPredictionsResponse))
  ),

  rest.get(`${REACT_APP_API_URL}/predictions/:predictionId`, (req, res, ctx) =>
    res.once(
      ctx.status(400),
      ctx.json({ error: "There was an error on the server" })
    )
  ),

  rest.get(`${REACT_APP_API_URL}/predictions/:predictionId`, (req, res, ctx) =>
    res(ctx.status(200), ctx.json(mockGetPredictionByIdResponse))
  ),

  rest.post(`${REACT_APP_API_URL}/predictions/create`, (req, res, ctx) =>
    res.once(
      ctx.status(400),
      ctx.json({ error: "There was an error on the server" })
    )
  ),

  rest.post(`${REACT_APP_API_URL}/predictions/create`, (req, res, ctx) =>
    res(ctx.status(201))
  ),

  rest.delete(
    `${REACT_APP_API_URL}/predictions/delete/:predictionId`,
    (req, res, ctx) =>
      res.once(
        ctx.status(404),
        ctx.json({ error: "There was an error on the server" })
      )
  ),

  rest.delete(
    `${REACT_APP_API_URL}/predictions/delete/:predictionId`,
    (req, res, ctx) => res(ctx.status(201))
  ),

  rest.patch(
    `${REACT_APP_API_URL}/predictions/update/:predictionId`,
    (req, res, ctx) =>
      res.once(
        ctx.status(404),
        ctx.json({ error: "There was an error on the server" })
      )
  ),

  rest.patch(
    `${REACT_APP_API_URL}/predictions/update/:predictionId`,
    (req, res, ctx) => res(ctx.status(200))
  ),
];
