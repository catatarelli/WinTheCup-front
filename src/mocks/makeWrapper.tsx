import React from "react";
import { Provider } from "react-redux";
import { store } from "../redux/store";

const makeWrapper = ({ children }: { children: JSX.Element }) => (
  <Provider store={store}>{children}</Provider>
);

export default makeWrapper;
