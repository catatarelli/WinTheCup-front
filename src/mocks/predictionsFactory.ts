/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-return */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import { Factory } from "fishery";
import { faker } from "@faker-js/faker";
import type { PredictionStructure } from "../redux/features/user/userTypes";

const predictionFactory = Factory.define<PredictionStructure>(() => ({
  match: "Argentina vs England",
  goalsTeam1: faker.datatype.number(),
  goalsTeam2: faker.datatype.number(),
  redCards: faker.datatype.number(),
  yellowCards: faker.datatype.number(),
  penalties: faker.datatype.number(),
  picture: faker.image.sports(),
  backupPicure: faker.image.sports(),
}));

export const getRandomPredictionsList = (number: number) =>
  predictionFactory.buildList(number);
