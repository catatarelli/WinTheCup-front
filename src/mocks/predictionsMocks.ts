import { type ImageSourcePropType } from "react-native";
import {
  type PredictionsState,
  type PredictionsResponse,
  type PredictionStructure,
} from "../redux/features/predictions/predictionsTypes";
import {
  getRandomPrediction,
  getRandomPredictionsList,
} from "./predictionsFactory";

export const mockgetPredictionsResponse: PredictionsResponse = {
  predictions: getRandomPredictionsList(10) as PredictionStructure[],
};

export const emptyPredictionsState = {
  predictions: [],
  currentPrediction: {
    match: "",
    goalsTeam1: 0,
    goalsTeam2: 0,
    redCards: 0,
    yellowCards: 0,
    penalties: 0,
    picture: "" as ImageSourcePropType,
    createdBy: "",
    id: "",
  },
};

export const currentPredictionMock: PredictionStructure = {
  match: "Argentina vs Mexico",
  goalsTeam1: 3,
  goalsTeam2: 2,
  createdBy: "543",
  id: "234",
  picture: "asdf" as ImageSourcePropType,
};

export const mockPredictionsState: PredictionsState = {
  predictions: getRandomPredictionsList(10) as PredictionStructure[],
  currentPrediction: currentPredictionMock,
};

export const mockGetPredictionByIdResponse: PredictionStructure = {
  ...(getRandomPrediction() as PredictionStructure),
  id: "56789",
};
