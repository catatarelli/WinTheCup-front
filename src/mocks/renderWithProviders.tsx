import React, { type PropsWithChildren } from "react";
import { render } from "@testing-library/react-native";
import type { RenderOptions } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { type store, type RootState } from "../redux/store";
import { uiReducer } from "../redux/features/ui/uiSlice";
import { userReducer } from "../redux/features/user/userSlice";
import { NavigationContainer } from "@react-navigation/native";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: typeof store;
}

export const renderWithProviders = (
  ui: React.ReactElement,
  {
    preloadedState,
    store = configureStore({
      reducer: { ui: uiReducer, user: userReducer },
      preloadedState,
    }),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({
    children,
  }: PropsWithChildren<Record<string, unknown>>): JSX.Element => (
    <>
      <Provider store={store}>
        <NavigationContainer>{children}</NavigationContainer>
      </Provider>
    </>
  );

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};
