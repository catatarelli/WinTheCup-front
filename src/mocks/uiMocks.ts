import type { UiState } from "../redux/features/uiSlice/uiTypes.js";

export const emptyModalMock: UiState = {
  modal: "",
  isError: false,
};

export const mockUiStateSuccessModal: UiState = {
  modal: "Account created successfully",
  isError: false,
};

export const mockUiStateErrorModal: UiState = {
  modal: "User is already registered",
  isError: true,
};
