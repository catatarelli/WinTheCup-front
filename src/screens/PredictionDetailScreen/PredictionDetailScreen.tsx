import { SafeAreaView, Text } from "react-native";
import React from "react";
import screenStyles from "../../styles/screen.styles";

import headingStyles from "../../styles/headings.styles";
import PredictionDetail from "../../components/PredictionDetail/PredictionDetail";
import { getRandomPrediction } from "../../mocks/predictionsFactory";

const prediction = getRandomPrediction();

const PredictionDetailScreen = (): JSX.Element => (
  <SafeAreaView style={screenStyles.container}>
    <Text style={headingStyles.pageTitle}>Prediction Detail</Text>
    <PredictionDetail prediction={prediction} />
  </SafeAreaView>
);

export default PredictionDetailScreen;
