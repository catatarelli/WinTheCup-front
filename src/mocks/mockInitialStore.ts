import { configureStore } from "@reduxjs/toolkit";
import { uiReducer } from "../redux/features/ui/uiSlice";
import { userReducer } from "../redux/features/user/userSlice";
import { store } from "../redux/store";
import { mockUiStateErrorModal, mockUiStateSuccessModal } from "./uiMocks";
import { loggedInUserStateMock } from "./userMocks";

export const mockInitialStoreSuccessModal: typeof store = configureStore({
  reducer: {
    ui: uiReducer,
    user: userReducer,
  },
  preloadedState: {
    ui: mockUiStateSuccessModal,
    user: loggedInUserStateMock,
  },
});

export const mockInitialStoreErrorModal: typeof store = configureStore({
  reducer: {
    ui: uiReducer,
    user: userReducer,
  },
  preloadedState: {
    ui: mockUiStateErrorModal,
  },
});
