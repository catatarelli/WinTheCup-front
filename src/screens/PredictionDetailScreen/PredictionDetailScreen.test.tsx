/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import React from "react";
import { screen, render } from "@testing-library/react-native";
import PredictionDetailScreen from "./PredictionDetailScreen";
import { Provider } from "react-redux";
import { mockInitialStoreSuccessModal } from "../../mocks/mockInitialStore";
import { NavigationContainer } from "@react-navigation/native";

beforeEach(() => {
  jest.clearAllMocks();
});

describe("Given a PredictionDetailScreen component", () => {
  describe("When it's rendered", () => {
    test("Then it should show the title 'Prediction Detail' and a card with a prediction", async () => {
      const titleText = "Prediction Detail";
      const predictionId = "predictionDetail";

      render(
        <Provider store={mockInitialStoreSuccessModal}>
          <NavigationContainer>
            <PredictionDetailScreen />
          </NavigationContainer>
        </Provider>
      );

      const title = screen.queryByText(titleText);
      const prediction = screen.queryByTestId(predictionId);

      expect(title).toBeDefined();
      expect(prediction).toBeDefined();
    });

    test("Then it should match the snapshot", () => {
      const screen = render(
        <Provider store={mockInitialStoreSuccessModal}>
          <NavigationContainer>
            <PredictionDetailScreen />
          </NavigationContainer>
        </Provider>
      );
      expect(screen).toMatchSnapshot();
    });
  });
});
