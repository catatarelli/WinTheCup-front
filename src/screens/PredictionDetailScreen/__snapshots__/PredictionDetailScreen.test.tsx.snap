// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a PredictionDetailScreen component When it's rendered Then it should match the snapshot 1`] = `
<RCTScrollView
  style={
    Object {
      "backgroundColor": "#fff",
      "height": "100%",
    }
  }
>
  <View>
    <RCTSafeAreaView
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "center",
          "justifyContent": "center",
        }
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#3F3F3F",
            "fontSize": 30,
            "marginBottom": 10,
            "marginTop": 50,
          }
        }
      >
        Prediction Detail
      </Text>
      <View
        style={
          Object {
            "backgroundColor": "#EEEEE4",
            "borderRadius": 15,
            "display": "flex",
            "elevation": 6,
            "height": 625,
            "justifyContent": "space-between",
            "margin": 10,
            "marginBottom": 50,
            "paddingBottom": 10,
            "shadowColor": "#000",
            "shadowOffset": Object {
              "height": 3,
              "width": 0,
            },
            "shadowOpacity": 0.27,
            "shadowRadius": 4.65,
            "width": 321,
          }
        }
        testID="predictionDetail"
      >
        <Image
          source={
            Object {
              "uri": undefined,
            }
          }
          style={
            Object {
              "borderTopLeftRadius": 15,
              "borderTopRightRadius": 15,
              "height": 164,
              "resizeMode": "cover",
              "width": 321,
              "zIndex": 100,
            }
          }
          testID="predictionPicture"
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "space-between",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 20,
                "marginBottom": 10,
              }
            }
          >
            Argentina vs Mexico
          </Text>
          <View
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "flexDirection": "row",
                "justifyContent": "space-around",
                "marginBottom": 10,
                "width": 210,
              }
            }
          >
            <Text
              style={
                Object {
                  "borderColor": "#3F3F3F",
                  "borderWidth": 1,
                  "fontSize": 20,
                  "height": 45,
                  "paddingHorizontal": 10,
                  "paddingVertical": 5,
                  "textAlign": "center",
                  "width": 45,
                }
              }
            >
              3
            </Text>
            <Text
              style={
                Object {
                  "fontSize": 20,
                }
              }
            />
            <Text
              style={
                Object {
                  "borderColor": "#3F3F3F",
                  "borderWidth": 1,
                  "fontSize": 20,
                  "height": 45,
                  "paddingHorizontal": 10,
                  "paddingVertical": 5,
                  "textAlign": "center",
                  "width": 45,
                }
              }
            >
              2
            </Text>
          </View>
          <View
            style={
              Object {
                "alignItems": "center",
                "display": "flex",
                "justifyContent": "space-around",
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 20,
                  "marginVertical": 5,
                }
              }
            >
              Red cards
            </Text>
            <Text
              style={
                Object {
                  "borderColor": "#3F3F3F",
                  "borderWidth": 1,
                  "fontSize": 20,
                  "height": 45,
                  "paddingHorizontal": 10,
                  "paddingVertical": 5,
                  "textAlign": "center",
                  "width": 45,
                }
              }
            />
            <Text
              style={
                Object {
                  "fontSize": 20,
                  "marginVertical": 5,
                }
              }
            >
              Yellow cards
            </Text>
            <Text
              style={
                Object {
                  "borderColor": "#3F3F3F",
                  "borderWidth": 1,
                  "fontSize": 20,
                  "height": 45,
                  "paddingHorizontal": 10,
                  "paddingVertical": 5,
                  "textAlign": "center",
                  "width": 45,
                }
              }
            />
            <Text
              style={
                Object {
                  "fontSize": 20,
                  "marginVertical": 5,
                }
              }
            >
              Penalties
            </Text>
            <Text
              style={
                Object {
                  "borderColor": "#3F3F3F",
                  "borderWidth": 1,
                  "fontSize": 20,
                  "height": 45,
                  "paddingHorizontal": 10,
                  "paddingVertical": 5,
                  "textAlign": "center",
                  "width": 45,
                }
              }
            />
          </View>
        </View>
        <View
          style={
            Object {
              "display": "flex",
              "flexDirection": "row",
              "justifyContent": "space-around",
            }
          }
        >
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#3F3F3F",
                "borderRadius": 20,
                "color": "#fff",
                "height": 57,
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 0,
                "marginVertical": 25,
                "opacity": 0.8,
                "padding": 6,
                "width": 120,
                "zIndex": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontSize": 20,
                }
              }
              testID="editButton"
            >
              Edit
            </Text>
          </View>
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#3F3F3F",
                "borderRadius": 20,
                "color": "#fff",
                "height": 57,
                "justifyContent": "center",
                "marginBottom": 10,
                "marginTop": 0,
                "marginVertical": 25,
                "opacity": 0.8,
                "padding": 6,
                "width": 120,
                "zIndex": 1,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontSize": 20,
                }
              }
              testID="deleteButton"
            >
              Delete
            </Text>
          </View>
        </View>
      </View>
    </RCTSafeAreaView>
  </View>
</RCTScrollView>
`;
